### ISSUE TITLE
Improve MAP Estimates

### ISSUE THREAD

**Issue by null-a** [probmods/webppl#140]
The way we compute the MAP in `makeMarginalERP` doesn't work very well when the marginal comes from a sampling algorithm. This is because we pick the sample with the highest count as the MAP value, but in large enough spaces all counts are 1.

We can do better than this by looking at each sample's score (where available) - `HashMH` and `IncrementalMH` already do this.

There are a couple of things we can improve:

1. `HashMH` and `IncrementalMH` overwrite the `MAP` *function* on the marginal ERP with an *object*, which makes the interface inconsistent.
2. More generally, we could try to refactor the MAP computation to reduce duplication etc. (Related: #64.)

**Comment by iffsid** [comment]
The reliance on counts is indeed unpleasant.
One way around this is to instrument the histogram creation with objects that not only record the value and count (called `prob` in the object), but also the score (appropriately incremented) for each trace/particle as it is added to the histogram.

The reason I put in the computations as functions is that the computation of _erp-specific_ values, like the MAP estimate or entropy, i.e., those that don't change w.r.t anything but the erp itself, can be easily memoised. 
The MAP and entropy methods implemented as prototypes for the ERP class/type compute the requisite values once and then modify the local method to just return that value.

**Comment by mhtess** [comment]
Possibly related. Would it be possible to implement a Highest Posterior Density interval estimator for ERPs? This would be tremendously useful for bayesian data analysis. 

Ideally it would take the arguments of the interval, defaulting to 95%.  e.g.. `myERP.HPD([0.025,0.975])`

**Comment by ngoodman** [comment]
sounds like a great thing for you to implement @mhtessler! ;)


-N

**Comment by null-a** [comment]
> We can do better than this by looking at each sample's score

I've realised that this isn't always the case, as doing so doesn't account for marginalization.

For example:

```javascript
function() {
  var a = flip();
  var b = flip();
  return a + b;
}
```

With this model all samples have the same score, so the MAP estimate (from `HashMH` say) will always be the first sample taken. This means the distribution on MAP estimates is the same as the marginal of the model which seems odd. (You may as well take a single sample and use that instead.) In this case then, the approach taken in `ERP.prototype.MAP` has some utility.

So both approaches are useful, but the best approach doesn't coincide with whether or not an MCMC method is used. We perhaps need a straight-forward interface to present both options to the user.

Thinking about this a little more, I wonder whether it's even reasonable to call the strategy of returning the (arg)max over sample scores a MAP estimate. I recall hearing the term 'most probable explanation' somewhere - is that applicable here?

**Comment by iffsid** [comment]
> so the MAP estimate (from HashMH say) will always be the first sample taken.

That's true. The choice of what is actually used can be left to an `aggregator` that's passed in separately. I believe it is the case that when the model returns `Real`s, one would construct the posterior itself from the scores rather than the counts (because with counts one would simply get the uniform distribution), and in that case, the MAP estimate is indeed the (arg)max over the sample scores.

We might be able to automate the aggregator based on a simple inspection of the type of the distribution being returned. If there is a `Real` involved, we use the score, else the count.

**Comment by stuhlmueller** [comment]
> Thinking about this a little more, I wonder whether it's even reasonable to call the strategy of returning the (arg)max over sample scores a MAP estimate. 

It depends on what distribution we are talking about. If we are talking about the marginal distribution (which seems most sensible to me), then it doesn't make sense to call the (arg)max over sample scores a MAP estimate. If we are talking about the joint distribution on all variables in the program, then it does make sense. This applies equally in the continuous and discrete setting. To see the difference, consider this model:

    var model = function(){
      var x = gaussian(1, 1);
      var y = gaussian(1, Math.abs(x));
      return x;
    }

The MAP based on (joint) scores is (close to) 0, whereas the MAP based on (marginal) samples is 1.

**Comment by null-a** [comment]
> HashMH and IncrementalMH overwrite the MAP function on the marginal ERP with an object, which makes the interface inconsistent.

This is no longer the case.

I think the main outstanding issue here now is the confusion caused by referring to two different ideas with MAP and onlyMAP. I'll rename the issue accordingly.

This isn't a fully fleshed out suggestion, but perhaps we could do something like:

* Rename `ERP.prototype.MAP` to something along the lines of `argmaxScore`. I think this makes it clearer that calling this on an ERP made from samples (drawn from a large space) is unlikely to be useful. Not mentioning "posteriori" is also sensible since this method can be called on ERP which don't represent a posterior distribution.
* Rename the MCMC `onlyMAP` option to something along the lines of "highest-scoring execution". (Andreas used that phrase [here](https://github.com/probmods/webppl/issues/284#issuecomment-170671767).) This better describes what you're getting back, and it might help avoid surprise when this and the MAP of the marginal don't coincide. (I'm assuming that most people would think that `onlyMAP` returns the maximum of the marginal.)

For future reference, it's also worth [noting](https://github.com/probmods/webppl/issues/284#issuecomment-170671767) Andreas' suggestion for combining the `onlyMAP` and `justSample` options.
